imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+        if secondary_vector_missing:
+            vector_index = secondary_count - secondary_vector_missing
+            secondary_end = (
+                secondary_boundaries[-1]
+                + uncompressed_vector_size * secondary_vector_missing
+            )
+            uncompressed_vectors = vector_bits[secondary_boundaries[-1] : secondary_end]
+            for i in range(0, len(uncompressed_vectors), uncompressed_vector_size):
+                decoded_vector = MagL1a.unpack_one_vector(
+                    uncompressed_vectors[i : i + uncompressed_vector_size],
+                    compression_width,
+                    False,
+                )
+                secondary_vectors[vector_index] = decoded_vector
+                secondary_vectors[vector_index][3] = first_secondary_vector[3]
+                vector_index += 1
+
+        # If there is a range data section, it describes all the data, compressed or
+        # uncompressed.
+        if has_range_data_section:
+            primary_vectors = MagL1a.process_range_data_section(
+                bit_array[end_vector : end_vector + (primary_count - 1) * 2],
+                primary_vectors,
+            )
+            secondary_vectors = MagL1a.process_range_data_section(
+                bit_array[
+                    end_vector + (primary_count - 1) * 2 : end_vector
+                    + (primary_count + secondary_count - 2) * 2
+                ],
+                secondary_vectors,
+            )
+
+        return primary_vectors, secondary_vectors
+
+    @staticmethod
+    def process_range_data_section(
+        range_data: np.ndarray, vectors: np.ndarray
+    ) -> np.ndarray:
+        """
+        Given a range data section and vectors, return an updated vector array.
+
+        Each range value has 2 bits. range_data will have a length of n*2, where n is
+        the number of vectors in vectors.
+
+        Parameters
+        ----------
+        range_data : numpy.ndarray
+            Array of range values, where each value is one bit.
+        vectors : numpy.ndarray
+            Array of vectors, where each vector is a tuple of (x, y, z, range).
+            The range value will be overwritten by range_data, and x, y, z will remain
+            the same.
+
+        Returns
+        -------
+        numpy.ndarray
+            Updated array of vectors, identical to vectors with the range values
+            updated from range_data.
+        """
+        if len(range_data) != (len(vectors) - 1) * 2:
+            raise ValueError(
+                "Incorrect length for range_data, there should be two bits per vector, "
+                "excluding the first."
+            )
+
+        updated_vectors: np.ndarray = np.copy(vectors)
+        range_str = "".join([str(i) for i in range_data])
+        for i in range(len(vectors) - 1):
+            range_int = int(range_str[i * 2 : i * 2 + 2], 2)
+            updated_vectors[i + 1][3] = range_int
```

Comment:
It bugged me how this went from a array of bits to a string to ints. Best i could do to simplify was like this into array of 2bit range values. Will be faster but not sure if it is any better?

```python
arr = np.array([0,0,0,1,1,0,1,1]) #[0,1,2,3] as 2 bit numbers in binary array
result = [(x<<1)+y for x,y in np.array_split(arr,len(arr)/2)]
```

Reply:
This was a slow bit in my profiling. But that change is less intuitive.... I'll contemplate it, thanks for the suggestion!